pkg/controller/linstornodeset/linstornodeset_controller.go

```diff
@@ -329,180 +335,301 @@ func (r *ReconcileLinstorNodeSet) reconcileResource(ctx context.Context, pns *pi
 	return nil
 }
 
-// This function is a mini-main function and has a lot of boilerplate code that doesn't make a lot of
-// sense to put elsewhere, so don't lint it for cyclomatic complexity.
-// nolint:gocyclo
-func (r *ReconcileLinstorNodeSet) reconcileSatNodes(pns *piraeusv1alpha1.LinstorNodeSet) []error {
-	pods := &corev1.PodList{}
-	labelSelector := labels.SelectorFromSet(pnsLabels(pns))
-	listOpts := []client.ListOption{
-		// Namespace: pns.Namespace, LabelSelector: labelSelector}
-		client.InNamespace(pns.Namespace), client.MatchingLabelsSelector{Selector: labelSelector},
-	}
-	err := r.client.List(context.TODO(), pods, listOpts...)
+func (r *ReconcileLinstorNodeSet) reconcileAllNodesOnController(ctx context.Context, nodeSet *piraeusv1alpha1.LinstorNodeSet) []error {
+	logger := logrus.WithFields(logrus.Fields{
+		"Name":      nodeSet.Name,
+		"Namespace": nodeSet.Namespace,
+		"Op":        "reconcileAllNodesOnController",
+	})
+	logger.Debug("start per-node reconciliation")
+
+	pods, err := r.getAllNodePods(ctx, nodeSet)
 	if err != nil {
 		return []error{err}
 	}
 
-	type satStat struct {
-		sat *piraeusv1alpha1.SatelliteStatus
-		err error
-	}
-
-	satelliteStatusIn := make(chan satStat)
-	satelliteStatusOut := make(chan error)
-
-	maxConcurrentNodes := 5
-	tokens := make(chan struct{}, maxConcurrentNodes)
-
-	var wg sync.WaitGroup
+	// Every node will send encountered errors or nil over this channel
+	output := make(chan error)
 
-	pns.Status.SatelliteStatuses = make([]*piraeusv1alpha1.SatelliteStatus, len(pods.Items))
+	for i := range pods {
+		pod := &pods[i]
+		// Registration can be done in parallel, so we handle per-node work in a separate go-routine
+		go func() {
+			output <- r.reconcilePod(ctx, nodeSet, pod)
+		}()
+	}
 
-	for i := range pods.Items {
-		wg.Add(1)
-		pod := pods.Items[i]
+	logger.Debug("start collecting per-node reconciliation results")
 
-		log := logrus.WithFields(logrus.Fields{
-			"podName":                        pod.Name,
-			"podNameSpace":                   pod.Namespace,
-			"podPhase":                       pod.Status.Phase,
-			"podNumber":                      i,
-			"maxConcurrentNodeRegistrations": maxConcurrentNodes,
-		})
-		log.Debug("NS reconcileSatNodes: reconciling node")
+	registerErrors := make([]error, 0)
 
-		sat := &piraeusv1alpha1.SatelliteStatus{
-			NodeStatus:          piraeusv1alpha1.NodeStatus{NodeName: pod.Spec.NodeName},
-			StoragePoolStatuses: make([]*piraeusv1alpha1.StoragePoolStatus, 0),
+	// Note: We collect over the _number_ of pods, not the pods it self. There is no guarantee on the order in which
+	// results for each pod are received.
+	for range pods {
+		err := <-output
```

Comment:
We now have stable status ordering, but the errors are still mixed. Could we order them too? Maybe a slice of channels, one for each pod, and then wait for each of them in turn?